lightning-persister/src/background_processor.rs

```diff
@@ -0,0 +1,219 @@
+use lightning::chain;
+use lightning::chain::chaininterface::{BroadcasterInterface, FeeEstimator};
+use lightning::chain::keysinterface::{ChannelKeys, KeysInterface};
+use lightning::ln::channelmanager;
+use lightning::ln::channelmanager::ChannelManager;
+use lightning::util::logger::Logger;
+use lightning::util::ser::Writeable;
+use std::sync::Arc;
+use std::sync::mpsc::{self, Sender, TryRecvError};
+use std::thread;
+use std::time::{Duration, SystemTime};
+
+/// BackgroundProcesser runs a background thread that monitors whether the ChannelManager it's given on
+/// start has new updates to persist. If it does, then it writes said ChannelManager to disk/backups.
+pub struct BackgroundProcessor {
+	#[cfg(test)]
+	pub thread_terminator: Sender<()>,
+	#[cfg(not(test))]
+	/// Used to send a message to the thread in BackgroundProcessor::start() to exit the thread.
+	thread_terminator: Sender<()>,
+}
+
+impl BackgroundProcessor {
+	/// Start the background thread that checks whether new updates are available for the ChannelManager,
+	/// and writes it to disk if so.
+	/// We also call ChannelManager::timer_chan_freshness_every_min(), since we're already eating the
+	/// cost of running a thread, we may as well pack as many operations into it as make sense.
+	// Mark as must_use because otherwise the result is dropped immediately, resulting in the thread
+	// being terminated.
+	#[must_use]
+	pub fn start<ChanSigner: 'static + ChannelKeys + Writeable, M: 'static + chain::Watch<Keys=ChanSigner>, T: 'static + BroadcasterInterface, K: 'static + KeysInterface<ChanKeySigner=ChanSigner>, F: 'static + FeeEstimator, L: 'static + Logger>(persister: Arc<dyn channelmanager::Persist<ChanSigner, M, T, K, F, L>>, manager: Arc<ChannelManager<ChanSigner, Arc<M>, Arc<T>, Arc<K>, Arc<F>, Arc<L>>>) -> Self {
+		let (sender, receiver) = mpsc::channel();
+		// See ChannelMananger::persistence_lock for more documentation on the persistence process.
+		thread::spawn(move || {
+			let mut current_time = SystemTime::now();
+			loop {
+				let updates_available = manager.block_until_needs_persist(Some(Duration::from_millis(1000)));
+				if updates_available {
+					if let Err(e) = persister.persist_manager(manager.clone()) {
+						println!("Errored persisting manager: {}", e);
```

Comment:
Do we have another alternative besides swallowing the error? What are the ramifications of failing to persist the `ChannelManager`? Perhaps we should document this at the trait level.

Reply:
Good point. Sans having some callback mechanism (which, I guess we could? But that seems really complex), I'm not sure what it could be. Luckily the ChannelMonitor persistence stuff will block us from doing anything that will result in funds loss if the hard drive dies, but, yes, we should at least document that users need to figure *something* out in their own persister implementation.

Reply:
FWIW, looks like you can pass back an error via `panic!` and retrieve it by joining on the thread handle.

https://doc.rust-lang.org/std/thread/struct.JoinHandle.html#method.join

Reply:
I presume that will not work if you have panic=abort set, which users may very well want to.

Reply:
Ah, more generally, a thread can have a return value which is wrapped in a `Result` as part of `join`'s return value. An error variant is returned when there is a panic.

Reply:
Ah, ok. Still, not sure its right, either - we want to be able to continue persisting after a temporary failure, too.

Reply:
Do you all think adding a basic callback function as a parameter could make sense? I added it, let me know what you think. @ariard you asked about this too.

Reply:
To close the loop on this, we discussed offline to always have the thread panic when the persister returns an error. This should be well-documented just as panics are documented in the Rust standard libraries.

With this approach, users have the flexibility of either handling the panic using `JoinHandle`, [configuring](https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/aborting-on-panic.html) to abort on panic, or writing a custom persister to handle the error without necessarily killing the thread.

@valentinewallace We'll need have `BackgroundProcessor` expose the `JoinHandle` to handle the first use case.

Reply:
Thanks, added this.